README.md
# Tornado

A Clojure library designed to generate CSS using [hiccup-like](https://github.com/weavejester/hiccup)
data structures with focus on simplicity.

## Changelog
is available [here](CHANGELOG.md)

## Benefits
- if you know Garden or Hiccup, nothing new can surprise you in Tornado
- works in both Clojure and ClojureScript
- the CSS data description is very simple - similar to Hiccup or Garden + everything can be found in a single namespace, you do not have to think about where to find what you need
- the code is easy to read, the compilation part is logical
- the library is very fast

## Usage


[![Clojars Project](https://img.shields.io/clojars/v/org.clojars.jansuran03/tornado.svg)](https://clojars.org/org.clojars.jansuran03/tornado)


### [Lein-tornado](https://github.com/JanSuran03/lein-tornado)

A hot-code reloading plugin for automatic compilation of Tornado stylesheets:

[![Clojars Project](https://img.shields.io/clojars/v/org.clojars.jansuran03/lein-tornado.svg)](https://clojars.org/org.clojars.jansuran03/lein-tornado)


If you are familiar with [garden](https://github.com/noprompt/garden), you should not have any problems with switching to Tornado.

Require a namespace ***tornado.core***, where you have available ***everything useful in this library***. Define some example CSS-hiccup like below and
run it with a function "repl-css" to see the compiled CSS:

```clojure
(require '[tornado.core :refer :all])
=> nil

(def styles
  [:#some-id {:width            (px 15)
              :color            :font-black
              :padding          (join 5 10 15 20)
              :gap              (join em 4 2)
              :background-color (rgb 100 150 200)}])
=> #'user/styles

(repl-css styles)

#some-id {
    width: 15px;
    color: #1A1B1F;
    padding: 5px 10px 15px 20px;
    gap: 4em 2em;
    background-color: rgb(100, 150, 200);
}

=> nil
```

For compiling **and saving** the stylesheet, there is a function tornado.compiler/css, also referred in tornado.core:

```clojure
;; First, you have to ensure that the path "resources/css" exists. In the plugin
;; Lein-tornado, this is not a problem anymore, the library creates the folders for you.
(css {:output-to "resources/css/example.css"} styles)
=> nil
```

ClojureScript:

```clojure
(ns example.core
  (:require [reagent.dom :as r-dom]
    [tornado.core :as t :refer [em]]) ; in ClojureScript, you cannot refer the whole namespace tornado.core
  (:require-macros [tornado.macros :refer [defunit]])) ; note that you cannot refer macros from tornado.core, like in Clojure

(defunit vw) ; (viewport width) compiles to "vw", string form of the reference function
(defunit celsius "°C") ; `celsius` is the reference, compiles to "°C"

;; Reagent does not support fully compilated params of every kind, but you can precompile :style.
(defn root []
  [:h2 {:style (compile-params {:color           :chocolate
                                :font-size       (em 15)
                                :position        :relative
                                :left            (vw 25)
                                :air-temperature (celsius 22)})}
   "Abc"])

(defn render []
  (r-dom/render
    [root]
    (.getElementById js/document "app"))
  (js/console.log "Hello World!"))
```

### The complete documentation with examples will be under this link: https://orgpad.com/s/SjH_TDbx4PH

## Plans for the future:

- Docs to CSS functions, better docs overall.
- More common utility functions (currently only 3).
- More at-rules functions (currently @media, @font-face, @keyframes).
- The code could be simpler on some places.
- More examples in the OrgPad document.

## Contact

Although there are more ways to contact me, you can send me an e-mail to **suran (dot) orgpad (at) gmail (dot) com**. I will
always try to reply as soon as possible. I am also on discord: **Honza_Suran#6703**

## License

Copyright © 2021 Jan Šuráň

Distributed under the [Eclipse Public License](#http://www.eclipse.org/legal/epl-2.0).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## 0.2.10
Fixed a very well hidden bug with at-media rules which caused the "not" keyword to be omitted.

Added an option useful for grid areas which allows to specify grid areas as namespaced keywords which allows to differentiate between similar grid areas: E.g.
```clojure
[:.foo {:display    :grid
        :grid-areas (grid-areas [(repeat 2 :header) (repeat 2 ::content) (repeat 2 :foo/ter)])}
  [:.bar {:grid-area ::content}]
  [:.footer {:grid-area :foo/ter}]
  [:.baz {:grid-area :header}]]
```
will compile into
```css
.foo {
    display: grid;
    grid-areas: "header header" "user-somens--content user-somens--content" "foo--ter foo--ter";
}

.foo .bar {
    grid-area: user-somens--content;
}

.foo .footer {
    grid-area: foo--ter;
}

.foo .baz {
    grid-area: header;
}
```

## 0.2.9
Added test namespaces (right at the beginning of patch 0.2.9) which will be run before every release and extended as new features come out.

A lot of defs and fully variadic defmacros in tornado.core were rewritten to defns and defmacros with stable arglists to enable arglist checks for Cursive
(a Clojure IDE plugin for IntelliJ IDEA) and maybe some more.

Some clj?s defs were rewritten to better match conventions by correctly differentiating between records and protocols.

BUGFIX: Fixed mixing of colors.

BUGFIX: At-font-face can now be placed inside the hiccup as well (could be placed only at the top until now).

NEW FEATURE: Enabled unnesting of seqs (created with e.g. `map` and `for`).

## 0.2.8
Added much more dynamism to the "join" common function:
```clojure
[[(em 3) (px 15) (fr 4) (px 3)]] ; with join:
(join (em 3) 15 (fr 4) 3)
```

## 0.2.7
If any unit has a zero value, it will be compiled to 0 instead of 0<unit-name>

## 0.2.6
Added a convenient function "join" which can be used for much simpler description of structures like
```clojure
[[(px 5) (px 10) (px 15) (px 20)]] ; with join:
(join 5 10 15 20)

[[(em 15) (em 20) (em 45)]] ; with join:
(join em 15 20 45)
```

## 0.2.5
**Bugfix**: Fixed hsl color compilation with pretty-print? set to false.

Functions on a few places replaced with dynamic bindings (*indent*, *compress?*).

## 0.2.4
Added a function `compile-params` for more convenient description of CSS styles in ClojureScript (Reagent).

Added functions `with-hue`, `with-saturation`, `with-lightness`, `with-alpha` you can use if you wanted to change e.g. only lightness of a color
(e.g. `(with-hue (rgb 50 100 150) 120)`, `(with-lightness "#FF0000" 0.3)`, `(with-alpha :black 0.5)`).

## 0.2.3
Fixed rgba color compilation. Added color compression to hex-format if `:pretty-print?` flag is set to false.

## 0.2.2
There was explicit java `Math` class on a few places - replaced with reader conditionals for java's `Math` and javascript's `js/Math` respectively.

Fixed mixing keyword, string or symbol colors which can be found in `tornado.colors/default-colors`.

## 0.2.1
`@rules` in one of the documentations replaced with at-rules to avoid collisions with the Closure compiler trying to parse it somehow.

# 0.2.0 ClojureScript support
Added support for ClojureScript. There are a few limitations that you have to require macros from a different file etc.

## 0.1.8
Removed duplicate selectors, added information to selectors in general.

## 0.1.7
Tiny improvements in code, removed a few printlines.

## 0.1.6
Some broken version, use **0.1.7** instead.

## 0.1.5
Small improvements in code.

## 0.1.4
Small improvements in code, updated README with plans for the future.

## 0.1.3
Added more flexibility to colors (e.g. `:font-black`, `:fontblack`, `"font-black"` `"fontblack"`, `'font-black`, `'fontblack`) can all represent the same color.

Removed an information in README about upcoming lein-tornado plugin - the plugin is already released.

## 0.1.2
Fixed a bug where colors such as rgb did not have the r, g, b values rounded properly (e.g. 154.837256 instead of 155).
It caused that the colors did not work.

## 0.1.1
Removed lein-tornado plugin dependency.

## 0.1.0
The initial release of the Tornado library.

src/tornado/macros.clj
(ns tornado.macros
  (:require [tornado.types])
  (:import (tornado.types CSSUnit CSSAttributeSelector CSSPseudoClass CSSPseudoClassFn
                          CSSFunction CSSCombinator CSSPseudoElement CSSAtRule)
           (clojure.lang PersistentList IFn)))

(defmacro defkeyframes
  "Defines a CSS @keyframes animation. The animation name should have a unique symbol
  for later reference to it and then animation frames in a format [progress params]:

  (defkeyframes fade-in-opacity
                [(u/percent 0) {:opacity 0}]
                [(u/percent 25) {:opacity 0.1}]
                [(u/percent 50) {:opacity 0.25}]
                [(u/percent 75) {:opacity 0.5}]
                [(u/percent 100) {:opacity 1}])

  Then, refer it the CSS hiccup list to make tornado compile it for later usage:

  (def styles
     (list
        fade-in-opacity
        ...))

  After that, you can assign this animation to whatever element you want:

  (def styles
     (list
        fade-in-opacity
        [:.some-element {:animation-duration (u/ms 500)
                         :animation-name     fade-in-opacity)}]
        [:#another-element {:animation-name  fade-in-opacity
                            :animation-delay (u/s 1.5)}]))

  With defkeyframes you can also define from & to progress animations:

  (defkeyframes translate-animation
                [:from {:transform (f/translate (u/px 100) (u/px 200)}]
                [:to {:transform (f/translate (u/px 200) (u/px 400)}])"
  [animation-name & frames]
  `(def ~animation-name (CSSAtRule. "keyframes" {:anim-name (str '~animation-name)
                                                 :frames    (list ~@frames)})))

(defmacro defunit
  "Creates a unit function which takes 1 argument and creates a CSSUnit record for future
   compilation. Defunit can take 1 arg: (defunit px)
                             or 2 args: (defunit percent \"%\").

   Usage of the defined units: (px 15)       ...  compiles to \"15px\"
                               (percent 33)  ...  compiles to \"33%\"

   CSSUnits can be added, subtracted, multiplied or divided by using function calc (you can
   also use these 4 keywords - they are defined just for better search in code:
   :add, :sub, :mul, :div

   E.g. (calc (px 500) :add 3 :mul (vw 5)) ... \"calc(500px + 3 * 5vw)\"."
  ([unit]
   `(defunit ~unit ~(str unit)))
  ([identifier css-unit]
   `(defn ~identifier [value#] (CSSUnit. ~css-unit value#))))

(defmacro defattributeselector
  "Attribute selectors select all descendant elements containing a given attribute,
  of which the value matches a given substring. All attribute selectors have
  different conditions for matching. Dashes count as words separators.
  By attributes, it is meant html attributes, e.g.:    div[class~=\"info\"]

  Usage:
     (defattributeselector contains-word \"~=\")
     => #'tornado.selectors/contains-word
     (contains-word :direction \"reverse\")
     => #tornado.types.CSSAttributeSelector{:compiles-to \"~=\"
                                            :attribute   :direction
                                            :subvalue    \"reverse\"}


  [attribute=\"value\"]:
     Selects all descendant elements which have a given parameter with a given value.
     In code: <has-val>
  - - - - - - - - - - - -
  With an html tag:
  a[attribute=\"value\"]:
     Selects all descendants of a html tag which have a given parameter
     with a given value.

  [attribute~=\"value\"]:
     Selects all descendant elements which have a given parameter with a value
     containing a given word (not substring, word).
     In code: <contains-word>

  [attribute|=\"value\"]:
     Selects all descendant elements which have a given parameter with a value
     starting with a given word (not substring, word).
     In code: <starts-with-word>

  [attribute^=\"value\"]:
     Selects all descendant elements which have a given parameter with a value
     starting with a given substring (unlike the \"|=\" selector, the substring
     does not have to be a whole word.
     In code: <starts-with>

  [attribute$=\"value\"]:
     Selects all descendant elements which have a given parameter with a value
     ending with a given substring. The substring does not have to be a whole word.
     In code: <ends-with>

  [attribute*=\"value\"]:
     Selects all descendant elements which have a given parameter with a value
     containing a given substring. (unlike the \"~=\" selector, the substring does
     not have to be a whole word).
     In code: <contains-subs>"
  [selector-name compiles-to]
  `(do (defn ~selector-name
         ([attr# subval#] (CSSAttributeSelector. ~compiles-to nil attr# subval#))
         ([tag# attr# subval#] (CSSAttributeSelector. ~compiles-to tag# attr# subval#)))
       (alter-meta! #'~selector-name assoc :arglists '([~'attribute ~'subvalue]
                                                       [~'tag ~'attribute ~'subvalue]))
       #'~selector-name))

(defmacro defpseudoclass
  "Defines a CSS pseudoclass. A CSS pseudoclass can activate some CSS properties on
  a css-class/css-id/html-element based on some current special state of the element.

  For example, hover: (defpseudoclass hover)
  When compiling a selectors sequence, e.g. [:.abc :#def hover], the resulting CSS
  selectors sequence will look like this: \".abc #def:hover\".

  So, what does it even do? We can give the element a special value on hover:
  ... [:a hover {:color :blue} ...] - when we hover over a link with our mouse, the
  text color of the link will turn blue until we put our mouse away.

  Defpseudoclass can also take 2 parameters, where the 2nd one will be the translation
  to CSS to avoid collisions between Clojure and CSS -
  e.g.(defpseudolass css-empty \"empty\")."
  ([pseudoclass]
   `(defpseudoclass ~pseudoclass ~(str pseudoclass)))
  ([identifier css-pseudoclass]
   `(def ~identifier (CSSPseudoClass. ~css-pseudoclass))))

(defmacro defpseudoclassfn
  "Creates a special CSS pseudoclass function, which compiles similarly as a standard
  CSS pseudoclass, but it is pseudoclass function with an argument.

  For example. if you wanted to only select every n-th argument:
  (defpseudoclassfn nth-child)
  (nth-child :odd)     ... compiles to   \"<parent>:nth-child(odd)\"
  (nth-child \"3n+1\")   ... compiles to   \"<parent>:nth-child(3n+1)\"

  Or if you wanted to show something based on the current language of the browser:
  (defpseudoclass lang)
  (lang \"en\") ... compiles to   \"<parent>:lang(en)\"

  To avoid collisions with some Clojure functions, you can give a second argument
  to defpseudoclassfn for a different translation to CSS:
  (defpseudoclass css-not \"not\")
  (css-not :p) ... compiles-to   \"not(p)\", which selects all descendants which are
  not a paragraph."
  ([pseudoclass]
   `(defpseudoclassfn ~pseudoclass ~(str pseudoclass)))
  ([pseudoclass compiles-to]
   `(defn ~pseudoclass [arg#] (CSSPseudoClassFn. ~compiles-to arg#))))

(defmacro defpseudoelement
  "Defines a CSS pseudoelement. A CSS pseudoelement activates some CSS properties on
  a special part of a css-class/css-id/html-element.

  For example, first-letter: (defpseudoclass first-letter)
  When compiling a selectors sequence, e.g. [:.abc :#def first-letter], the resulting CSS
  selectors sequence will look like this: \".abc #def::first-letter\".

  So, what does it even do? We can give the first letter of an element a special value:
  ... [:.abc :p first-letter {:font-size (u/px 60)} ...] - this causes the first letter
  of every paragraph in an element with class .abc to have the first letter significantly
  bigger than the rest of the paragraph."
  ([pseudoelement]
   `(def ~pseudoelement (CSSPseudoElement. ~(str pseudoelement)))))

(defmacro defcombinatorselector
  "Defines a combinator selector function which describes relationships between its
  arguments depending on the selector type:

  :#abc :.def is the default combinator selector - descendant selector. Affects all
  children with a class .def.

  child-selector \">\": is active when the given selectors are every of them a direct
  child of the previous one.

  adjacent-sibling (selector) \"+\": is active when the given html blocks elements or
  elements with a given class/id connected with the \"+\" sign are adjacent siblings.

  general-sibling (selector) \"~\" is active when the given selectors are on the same
  level of nesting; they do not have to be adjacent necessarily.

  Usage: [:.abc
           [:.def (child-selector :p :#ghi)]]
  compiles to   \".abc .def, .abc > p > #ghi\""
  [selector-name compiles-to]
  `(defn ~selector-name [& children#] (CSSCombinator. ~compiles-to children#)))

(defmacro defcssfn
  "Defines a CSS function. In most cases, you do NOT need to define a special compile-fn
  function - it should always be enough to use one of single-arg, space-join, comma-join.
  All of them compile the params, but: Single-arg gives you a warning if you give it more
  than 1 argument and compiles the args like comma-join. Comma-join compiles all its args
  and str/joins them with a comma. Space-join compiles all its args and str/joins them
  with a space. All these function also take the compiles-to argument and put it in front
  of a bracket enclosing the str/joined arguments.
  You can give this function 1, 2 or 3 arguments:

  (defcssfn translate)   (the default compile-fn is comma-join)
  (translate (u/px 80) (u/css-rem 6))   ... compiles to    \"translate(80px, 6rem)\"

  (defcssfn css-min \"min\")
  (css-min (u/px 500) (u/vw 40) (u/cm 20))   ... compiles to   \"min(500px, 40vw, 20cm)\"

  (defcssfn calc space-join)
  (calc (u/px 200) add 3 mul (u/percent 20))   ... compiles to   \"calc(200px + 3 * 20%)\"

  The arity(3) can be used like this to combine both previous features of the arity(2):
  (defcssfn my-clj-fn \"css-fn\" space-join)
  (my-clj-fn (u/s 20) (u/ms 500))   ... compiles to   \"css-fn(20s 500ms)\""
  ([fn-name]
   `(defcssfn ~fn-name ~(str fn-name) nil))
  ([fn-name css-fn-or-fn-tail]
   (condp instance? css-fn-or-fn-tail String `(defcssfn ~fn-name ~css-fn-or-fn-tail nil)
                                      PersistentList `(defcssfn ~fn-name ~(str fn-name) ~css-fn-or-fn-tail)
                                      IFn `(defcssfn ~fn-name ~(str fn-name) ~css-fn-or-fn-tail)
                                      (throw (IllegalArgumentException.
                                               (str "Error defining a CSS function " fn-name " with arity(2):"
                                                    "\nThe second argument " css-fn-or-fn-tail " is"
                                                    " neither a string nor a function.")))))
  ([clojure-fn-name compiles-to compile-fn]
   `(defn ~clojure-fn-name [& args#] (CSSFunction. ~compiles-to ~compile-fn args#))))

(defmacro cartesian-product
  "Given any number of seqs, this function returns a lazy sequence of all possible
  combinations of taking 1 element from each of the input sequences."
  [& seqs]
  (let [with-bindings (map #(vector (gensym) %) seqs)
        binding-symbols (mapv first with-bindings)
        for-bindings (vec (apply concat with-bindings))]
    `(for ~for-bindings ~binding-symbols)))